plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenShiftProjectCache.java

```diff
@@ -50,79 +52,119 @@ public OpenShiftProjectCache() {
 	}
 	
 	@Override
-	public synchronized Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
+	public Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
 		String key = getCacheKey(conn);
-		if(!cache.containsKey(key)) {
+		Set<IProjectAdapter> adapters = null;
```

Comment:
Most safe is 
Set<IProjectAdapter> adapters = getEntry(key);
if(adapters == null) {
   ...
}
Otherwise, in a moment between contains(key) and adapters = getEntry(key), entry may disappear. 
This is not very probable, but the one thing which is a bit more probable is that this entry may be removed while projects are iterated and listeners are notified of an adapter added.